Mathlib/Data/Real/Pi/Irrational.lean

```diff
@@ -0,0 +1,248 @@
+/-
+Copyright (c) 2022 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+import Mathlib.Analysis.SpecialFunctions.Integrals
+import Mathlib.Data.Real.Irrational
+
+/-!
+# `Real.pi` is irrational
+
+The main result of this file is `irrational_pi`.
+-/
+
+noncomputable section
+
+open intervalIntegral MeasureTheory.MeasureSpace Set Polynomial Real
+open scoped Nat
+
+private def I (n : ℕ) (θ : ℝ) : ℝ := ∫ x in (-1)..1, (1 - x ^ 2) ^ n * cos (x * θ)
+
+variable {n : ℕ} {θ : ℝ}
+
+private lemma I_zero : I 0 θ * θ = 2 * sin θ := by
+  rw [mul_comm, I]
+  simp [mul_integral_comp_mul_right, two_mul]
+
+private lemma recursion' (n : ℕ) :
+    I (n + 1) θ * θ ^ 2 = - (2 * 2 * ((n + 1) * (0 ^ n * cos θ))) +
+      2 * (n + 1) * (2 * n + 1) * I n θ - 4 * (n + 1) * n * I (n - 1) θ := by
+  rw [I]
+  let f (x : ℝ) : ℝ := 1 - x ^ 2
+  let u₁ (x : ℝ) : ℝ := f x ^ (n + 1)
+  let u₁' (x : ℝ) : ℝ := - (2 * (n + 1) * x * f x ^ n)
+  let v₁ (x : ℝ) : ℝ := sin (x * θ)
+  let v₁' (x : ℝ) : ℝ := cos (x * θ) * θ
+  let u₂ (x : ℝ) : ℝ := x * (f x) ^ n
+  let u₂' (x : ℝ) : ℝ := (f x) ^ n - 2 * n * x ^ 2 * (f x) ^ (n - 1)
+  let v₂ (x : ℝ) : ℝ := cos (x * θ)
+  let v₂' (x : ℝ) : ℝ := -sin (x * θ) * θ
+  have hfd : Continuous f := continuous_const.sub (continuous_pow 2)
+  have hu₁d : Continuous u₁' := ((continuous_mul_left _).mul (hfd.pow _)).neg
+  have hv₁d : Continuous v₁' := (continuous_cos.comp (continuous_mul_right _)).mul continuous_const
+  have hu₂d : Continuous u₂' :=
+    (hfd.pow _).sub ((continuous_const.mul (continuous_pow 2)).mul (hfd.pow _))
+  have hv₂d : Continuous v₂' :=
+    (continuous_sin.comp (continuous_mul_right _)).neg.mul continuous_const
+  have hu₁_eval_one : u₁ 1 = 0 := by simp only [u₁, f]; simp
+  have hu₁_eval_neg_one : u₁ (-1) = 0 := by simp only [u₁, f]; simp
+  have t : u₂ 1 * v₂ 1 - u₂ (-1) * v₂ (-1) = 2 * (0 ^ n * cos θ) := by
+    simp only [u₂, v₂, f, one_pow, sub_self, one_mul, neg_one_sq, neg_mul, cos_neg, sub_neg_eq_add]
+    simp only [← two_mul]
+  have hf : ∀ x, HasDerivAt f (- 2 * x) x := by
```

Comment:
I'm not sure if fun_prop is meant to solve these already, or that was just an in-progress PR...

But at least, can you move the x into the `have`, to avoid the first `intro x` line? The same applies for a few more have's below.

Reply:
Thanks, I've cleaned up the `intro`s